uhd: rfnoc: Add GRC binding and support for replay block

gmeyerho opened the pull request:
Creating initial review in my fork to silo this to UHD reviewers since this block has more new functionality than just adding bindings for the block. I am using integer message passing, the integer to command mapping is included in the block documentation and in the example. I created a user story to add a more interesting replay block example that may provide more benefit to the customer, but I wanted to make sure to get to the other RFNoC blocks first before circling back on that. But this example is nice because it is usable with the base image on at least the n3xx and x410 with streaming endpoints directly connected to the replay block.

## Description
Adds support for using the RFNoC replay block and a replay command block that is used to control the replay block

## Related Issue

## Which blocks/areas does this affect?
gr-uhd

## Testing Done
Ran included example. Video shows front panel of example with port 0 connected to cosine wave, port 1 connected to square wave. Record on port 0 for one offset and port 1 for another. Playback from previously used record offsets

https://user-images.githubusercontent.com/68391340/224850755-a369d6c5-89cc-491e-a36f-0cd125871d84.mp4



## Checklist
<!--- Go over all the following points, and put an `x` in all the
<!--- boxes that apply. Note that some of these may not be valid -->
<!--- for all PRs. -->

- [x] I have read the [CONTRIBUTING document](https://github.com/gnuradio/gnuradio/blob/main/CONTRIBUTING.md).
- [x] I have squashed my commits to have one significant change per commit. 
- [x] I [have signed my commits before making this PR](https://github.com/gnuradio/gnuradio/blob/main/CONTRIBUTING.md#dco-signed)
- [x] My code follows the code style of this project. See [GREP1.md](https://github.com/gnuradio/greps/blob/main/grep-0001-coding-guidelines.md).
- [ ] I have updated [the documentation](https://wiki.gnuradio.org/index.php/Main_Page#Documentation) where necessary.
- [ ] I have added tests to cover my changes, and all previous tests pass.

mbr0wn committed:
uhd: rfnoc: Add GRC binding and support for Vector IIR block

Signed-off-by: Grant Meyerhoff <grant.meyerhoff@ni.com>

Co-authored-by: Martin Braun <martin.braun@ettus.com>
Co-authored-by: Grant Meyerhoff <grant.meyerhoff@ni.com>

gmeyerho committed:
uhd: rfnoc: Add GRC binding and support for replay block

Signed-off-by: Grant Meyerhoff <grant.meyerhoff@ni.com>

gmeyerho committed:
uhd: rfnoc: fix literal_eval error

literal_eval() was used in the RFNoC graph GRC binding to sanitize
inputs. However, it was throwing an exception on drag and drop because
it requires a string containing a literal (e.g. string) rather than
just a string itself,

This was only an issue on drag/drop as the block code later does this transformation itself, but not before the first var_make()

However, the literal_eval() is not actually necessary, since the objects
passed in are of type TemplateArg, not just
string-or-something-that-evaluates-to-string, and thus we can use
facilities from TemplateArg to properly escape the values before usage.

Signed-off-by: Grant Meyerhoff <grant.meyerhoff@ni.com>
Co-authored-by: Martin Braun <martin.braun@ettus.com>
Signed-off-by: Martin Braun <martin.braun@ettus.com>

mbr0wn commented on gr-uhd/lib/usrp_block_impl.cc:
```diff
@@ -427,6 +427,7 @@ void usrp_block_impl::setup_rpc()
 
 void usrp_block_impl::msg_handler_command(pmt::pmt_t msg)
 {
+    d_debug_logger->warn("Here");
```
Don't forget to remove this before opening the official PR.

mbr0wn commented on gr-uhd/include/gnuradio/uhd/rfnoc_replay.h:
```diff
@@ -0,0 +1,157 @@
+    virtual void set_play_time(const ::uhd::time_spec_t& time) = 0;
+
+private:
+    virtual void _command_handler(const pmt::pmt_t& msg) = 0;
```
I think you can drop this from the public interface.

mbr0wn commented on gr-uhd/lib/rfnoc_replay_impl.cc:
```diff
@@ -0,0 +1,138 @@
+
+void rfnoc_replay_impl::_command_handler(const pmt::pmt_t& msg)
+{
+    if (pmt::is_pair(msg) && pmt::is_integer(pmt::cdr(msg))) {
```
This is too hacky. Please come up with a dictionary style for sending commands.

The message handling in `usrp_block` might have some ideas you can re-use. Commands could look like this maybe:
```py
{
    'command': 'start_continuous',
    'port': 0, # or 'chan'?
    'num_samps': 1234, # if applicable
    'time': (100, 0.1234), # see usrp_block for this
}
```

gmeyerho replied:
I didn't see a command option block available to nicely produce a dictionary message, so I created a block to create these replay dictionary messages based on the qt-gui msgpushbutton. I considered a more generic dictionary message block, but figured this gave better usability for our specific use-case.

mbr0wn replied:
You can simply convert Python dicts to PMTs using `pmt.to_pmt()`, so this probably wasn't strictly necessary, but then again this is such a common use case for the replay block that I think it's a good addition.

mbr0wn commented on gr-uhd/grc/gen_rfnoc_replay_block_yml.py:
```diff
@@ -0,0 +1,221 @@
+  hide: ${'$'}{ 'part' if num_out_ports > ${n} else 'all' }
+"""
+
+INPUT_PARAM = """- domain: rfnoc
```
Can we make it such that the ports are labeled in0, in1, etc.? When I opened this, they were all called `in`.

mbr0wn commented on gr-uhd/lib/rfnoc_replay_impl.h:
```diff
@@ -0,0 +1,56 @@
+/* -*- c++ -*- */
+/*
+ * Copyright 2019 Ettus Research, a National Instruments Brand.
+ * Copyright 2020 Free Software Foundation, Inc.
```
You can drop the FSF lines in this PR.

mbr0wn commented on gr-uhd/include/gnuradio/uhd/rfnoc_replay.h:
```diff
@@ -0,0 +1,157 @@
+     *
+     * \param port Which output port of the replay block to use
+     */
+    virtual void set_record_port(const size_t port) = 0;
+
+    /*! Sets the port for the next playback
+     *
+     * \param port Which output port of the replay block to use
+     */
+    virtual void set_play_port(const size_t port) = 0;
```
I don't think you need the play port, record port, or play time APIs at all. They can all be passed in atomically with the command. This adds a statefulness that a) is not necessary and b) I really didn't understand it for the first minutes of reviewing this.

However, I would suggest adding an `issue_stream_command()` API as a substitute for the command handler API. You could potentially even use that from the command handler.

mbr0wn replied:
Another way to view this: With this API, you have to first set the port (through the Python API) and then send a command (through the command API). That's a bit confusing.

mbr0wn commented on gr-uhd/lib/rfnoc_replay_impl.cc:
```diff
@@ -0,0 +1,138 @@
+#include "rfnoc_replay_impl.h"
+#include <gnuradio/io_signature.h>
+#include <uhd/version.hpp>
+#include <volk/volk.h>
```
Spurious include? Or are we using this somewhere

mbr0wn reviewed (commented):
I LOOOOVE  :heart: this, especially the example. My main complaints are about the command format and the unnecessary state, and both have the same solution.

daniestevez committed:
modtool: fix formatting of python qa template

Two spaces were missing from an indented import statement in the
template for Python QA tests.

Signed-off-by: Daniel Estévez <daniel@destevez.net>

haru-02 committed:
grc: make node states consistant

Signed-off-by: haru-02 <rahulbalaji78@gmail.com>

haru-02 committed:
grc: prevent Generate/Run silent fails for unsaved GRC flowgraphs

Signed-off-by: haru-02 <rahulbalaji78@gmail.com>

mbr0wn commented on gr-uhd/lib/rfnoc_replay_impl.cc:
```diff
@@ -0,0 +1,201 @@
+        return;
+    }
+    if (command == "get_record_fullness") {
+        d_logger->debug("Record fullness of port {:d}: {:d}",
+                        cmd_port,
+                        d_wrapped_ref->get_record_fullness(cmd_port));
```
This is a nice addition -- but no good deed (or good idea) goes unpunished. This could be even more useful if you post the fullness on an output message port instead of logging it. For those who want to log it, no problem, they can simply connect the output message port to a message logger. But this would allow other blocks to programmatically query the fullness, and maybe act upon that.

mbr0wn commented on gr-uhd/python/uhd/replaymsgpushbutton.py:
```diff
@@ -0,0 +1,70 @@
+
+        self.lbl = lbl
+
+        self.replayDict = pmt.dict_add(pmt.make_dict(), pmt.intern('command'), pmt.intern(command))
```
You can also simply make `self.replayDict` a Python dictionary, which makes the code much more readable.

mbr0wn commented on gr-uhd/python/uhd/replaymsgpushbutton.py:
```diff
@@ -0,0 +1,70 @@
+        self.replayDict = pmt.dict_add(self.replayDict, pmt.intern('time'), pmt.from_float(time))
+
+    def onBtnClicked(self, pressed):
+        self.message_port_pub(pmt.intern("pressed"), self.replayDict)
```
If you want to make it a Python dict, then you need to change this line to
```suggestion
        self.message_port_pub(pmt.intern("pressed"), pmt.to_pmt(self.replayDict))
```

mbr0wn reviewed (approved):
All my comments are pretty minor.

mbr0wn commented on gr-uhd/lib/rfnoc_replay_impl.cc:
```diff
@@ -0,0 +1,201 @@
+        play(cmd_offset, cmd_size, cmd_port, cmd_time, false);
+        return;
+    }
+    if (command == "play_continuous") {
```
I'm thinking it might be more consistent if you make the `repeat` argument part of the command dict. Or you keep the `play_continuous` command, but allow the `play` command to carry a `repeat` argument which defaults to `false`. Or none of the above... I don't have a definitive opinion. Given your move to using a command dict, the `repeat` argument seems consistent though.

mbr0wn commented:
@gmeyerho This can be closed now right?
